nemo/lightning/megatron_parallel.py

```diff
@@ -43,15 +45,43 @@ def convert_output(self, output: torch.Tensor) -> torch.Tensor: ...
 
 
 def default_data_step(dataloader_iter: Iterator[DataT]) -> DataT:
-    batch = next(dataloader_iter)
+    """
+    Moves the data to a device.
+
+    In this case we utilize the match function to unpack the dataloader iterator. There may be a wrapper on the dataloader
+    iter from here: https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/fabric/strategies.py#L441.
 
-    if isinstance(batch, tuple) and len(batch) == 3:
-        batch = batch[0]
+    This will not subset the data for your with context parallel so please override this function if you
+    want to use context parallel.
 
-    if isinstance(batch, dict):
-        batch = {k: v.cuda(non_blocking=True) for k, v in batch.items()}
+    Examples:
+        If the dataloader_iter returns: [Tuple[<tensor>, <int>, <int>]] -> move to device
+        If the dataloader_iter returns: [<tensor>, <tensor>] -> move to device
 
-    return batch
+    Returns:
+        DataT: The data moved to the device.
+    """
+    if parallel_state.get_context_parallel_world_size() > 1:
+        raise ValueError(
+            "Default data step is being used in a context parallel environment."
+            "Please define your own data step that appropriately slices the data for context parallel."
+        )
+
+    match next(dataloader_iter):
+        # If its wrapped in a tuple, unpack it.
+        case (batch, int(_), int(_)):
+            pass
+        # Canonical case.
+        case batch:
+            pass
+        # If the dataloader_iter is empty, return None.
```

Comment:
This comment is not what happens anymore. Now in your `else` you say `raise ValueError(...):`. Are we sure we don't want to pass through a None on the iter? Otherwise we should prob change line 79 or 84 to `return None` and drop the `ValueError` piece.

Reply:
Actually what I hope would happen is that the `match(dataloader_iter)` would trigger an `StopIteration` or whatever it is exception rather than returning None. The StopIteration exception ideally would raise through this function (what happens now on the match fail) and be handled elsewhere. So I _think_ your case _ should never get hit...

Reply:
Maybe a better error message would be ``None` batch returned at the end of iteration rather than a `StopIteration` exception being raised, why?`

Reply:
So I only added this to be past the linter. I updated the comment so that its more accurate. My only concern would be how it plays out with drop_last=True since maybe that would dump some None's?